aws_signing_helper/update.go

```diff
@@ -0,0 +1,205 @@
+package aws_signing_helper
+
+import (
+	"bufio"
+	"log"
+	"math"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+)
+
+const UpdateRefreshTime = time.Minute * time.Duration(5)
+const AwsSharedCredentialsFileEnvVarName = "AWS_SHARED_CREDENTIALS_FILE"
+
+// Structure to contain a temporary credential
+type TemporaryCredential struct {
+	AccessKeyId     string
+	SecretAccessKey string
+	SessionToken    string
+	Expiration      time.Time
+}
+
+// Updates credentials in the credentials file for the specified profile
+func Update(credentialsOptions CredentialsOpts, profile string, once bool) {
+	var refreshableCred = TemporaryCredential{}
+	var nextRefreshTime time.Time
+	for {
+		credentialProcessOutput, err := GenerateCredentials(&credentialsOptions)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// Assign credential values
+		refreshableCred.AccessKeyId = credentialProcessOutput.AccessKeyId
+		refreshableCred.SecretAccessKey = credentialProcessOutput.SecretAccessKey
+		refreshableCred.SessionToken = credentialProcessOutput.SessionToken // nosemgrep
+		refreshableCred.Expiration, _ = time.Parse(time.RFC3339, credentialProcessOutput.Expiration)
+		if (refreshableCred == TemporaryCredential{}) {
+			log.Println("no credentials created")
+			syscall.Exit(1)
+		}
+
+		// Get credentials file contents
+		lines, err := GetCredentialsFileContents()
+		if err != nil {
+			log.Println("unable to get credentials file contents")
+			syscall.Exit(1)
+		}
+
+		// Write to credentials file
+		err = WriteTo(profile, lines, &refreshableCred)
+		if err != nil {
+			log.Println("unable to write to AWS credentials file")
+			syscall.Exit(1)
+		}
+
+		if once {
+			break
+		}
+		nextRefreshTime = refreshableCred.Expiration.Add(-UpdateRefreshTime)
+		log.Println("Credentials will be refreshed at", nextRefreshTime.String())
+		time.Sleep(time.Until(nextRefreshTime))
+	}
+}
+
+// Assume that the credentials file is located in the default path: `~/.aws/credentials`
+func GetCredentialsFileContents() ([]string, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		log.Println("unable to locate the home directory")
+		return nil, err
+	}
+
+	awsCredentialsPath := os.Getenv(AwsSharedCredentialsFileEnvVarName)
+	if awsCredentialsPath == "" {
+		awsCredentialsPath = filepath.Join(homeDir, ".aws", "credentials")
+	}
+	if err = os.MkdirAll(filepath.Dir(awsCredentialsPath), 0600); err != nil {
+		log.Println("unable to create credentials file")
+		return nil, err
+	}
+
+	readOnlyCredentialsFile, err := os.OpenFile(awsCredentialsPath, os.O_RDONLY|os.O_CREATE, 0600)
+	if err != nil {
+		log.Println("unable to get or create read-only AWS credentials file")
+		syscall.Exit(1)
+	}
+	defer readOnlyCredentialsFile.Close()
```

Comment:
Style-wise, I think this would need to be before the `if` block. In this case, the whole process will exit, so it's not important. But in cases where you just return early, the `Close()` method would not be called at all.

Reply:
I don't think there are any other cases in which we return early (at least between the opening of the file at this line that closes it)? I think the reason we should do it this way is because the file may not be able to be opened at all. In that case, there would be an error that we could check and exit appropriately. Since the file wasn't opened in that case, we wouldn't need to close the file.